src/permissionSelectors.js

```diff
@@ -0,0 +1,37 @@
+/* @flow strict-local */
+import type { PerAccountState } from './types';
+// import { ensureUnreachable } from './types'; // (see commented callsite)
+import { Role, CreateWebPublicStreamPolicy } from './api/permissionsTypes';
+import { getRealm, getOwnUser } from './selectors';
+
+export function getCanCreateWebPublicStreams(state: PerAccountState): boolean {
```

Comment:
Is there a particular piece of the web code that this is modeled on, and/or is meant to have the same behavior as? It'd be helpful to point to it if so.

Reply:
Ah, right, yeah. So I did look at `user_can_create_web_public_streams` in `settings_data`—

```typescript
export function user_can_create_web_public_streams(): boolean {
    if (
        !page_params.server_web_public_streams_enabled ||
        !page_params.realm_enable_spectator_access
    ) {
        return false;
    }

    return user_has_permission(page_params.realm_create_web_public_stream_policy);
}
```

```typescript
function user_has_permission(policy_value: number): boolean {
    /* At present, nobody and by_owners_only is not present in
     * common_policy_values, but we include a check for it here,
     * so that code using create_web_public_stream_policy_values
     * or other supersets can use this function. */
    if (policy_value === settings_config.create_web_public_stream_policy_values.nobody.code) {
        return false;
    }

    if (page_params.is_owner) {
        return true;
    }

    if (
        policy_value === settings_config.create_web_public_stream_policy_values.by_owners_only.code
    ) {
        return false;
    }

    if (page_params.is_admin) {
        return true;
    }

    if (page_params.is_spectator) {
        return false;
    }

    /* At present, by_everyone is not present in common_policy_values,
     * but we include a check for it here, so that code using
     * common_message_policy_values or other supersets can use this function. */
    if (policy_value === settings_config.common_message_policy_values.by_everyone.code) {
        return true;
    }

    if (page_params.is_guest) {
        return false;
    }

    if (policy_value === settings_config.common_policy_values.by_admins_only.code) {
        return false;
    }

    if (page_params.is_moderator) {
        return true;
    }

    if (policy_value === settings_config.common_policy_values.by_moderators_only.code) {
        return false;
    }

    if (policy_value === settings_config.common_policy_values.by_members.code) {
        return true;
    }

    const current_datetime = new Date();
    const person_date_joined = new Date(user_join_date);
    const user_join_days =
        (current_datetime.getTime() - person_date_joined.getTime()) / 1000 / 86400;
    return user_join_days >= page_params.realm_waiting_period_threshold;
}
```

—but I found this way (the one in the branch) to be less awkward than a direct translation of that. For a few reasons:

1. A direct translation of `user_has_permission` can't really work if we want Flow to help us confirm that we've handled all possible `CreateWebPublicStreamPolicy` values. I know there's a bug with the `ensureUnreachable`, discussed upthread, but a fix or workaround will still probably depend on the switch-case structure in the branch.
2. `user_has_permission` is shared among other `user_can_` functions, and it has what would be dead code for us. For example, it checks if the the policy value is `settings_config.common_message_policy_values.by_everyone.code`. That's 5, and 5 is not even a value that the create-web-public-stream policy can take. (Because of the [product decision](https://zulip.com/help/configure-who-can-create-streams) that it's too risky to even let org owners permit anyone to create web-public streams.) The bit at the end, where it makes a `new Date()`, would also be dead code, since all of this policy's possible values are exhausted before then. For dead code, it seems pretty effortful to keep that code lying around such that it'd behave correctly if we ever actually invoke it (managing React re-rendering, etc.).

Then, I figured that clients are free to write code in a way that makes sense, as long as they respect the documented API that they're using. I believe I've done so, and I didn't find anything unexpected or warty in the API from reading this web-app code.

Reply:
For the record, I do think we've got the same behavior as the server's `user_can_create_web_public_streams` (modulo [some](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/event.20with.20.60create_web_public_stream_policy.60.3F/near/1381274) [inputs](https://github.com/zulip/zulip-mobile/pull/5384#issue-1238175522) aren't live-updating yet). But I'm also looking to the API documentation (what the different values mean for this policy, what the different roles are, etc.), and I'm glad to be aligning with that too.

Reply:
Cool, that reasoning makes sense.

I think it's still helpful to point at that web implementation just for comparison, or to say it's loosely based on / simplified from that one.

If you're thinking of the API docs as the primary thing this code is written against (which is good! better that than an implementation, as long as the docs are clear enough), then a link to the relevant bit of API docs would also be helpful.